js/store.js

```diff
@@ -296,58 +300,64 @@ const store = {
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      isNewFirstParty = true;
     }
 
     // add link in first party if it doesn't exist yet
     // and the third party is visible (i.e. not whitelisted)
-    if (!('thirdPartyHostnames' in  firstParty)
-      || !firstParty['thirdPartyHostnames'].includes(target)) {
-      if (!('isVisible' in thirdParty) || !thirdParty['isVisible']) {
-        if (!(this.onWhiteList(origin, target))) {
-          // show third party
-          thirdParty['isVisible'] = true;
-          isNewThirdParty = true;
-        } else {
+    if (!this.isFirstPartyConnectedToThirdParty(firstParty, target)) {
+      if (!this.isVisibleThirdParty(thirdParty)) {
+        if (this.onWhiteList(origin, target)) {
           // hide third party
           thirdParty['isVisible'] = false;
+        } else {
+          // show third party and all its connections
+          thirdParty['isVisible'] = true;
+          isNewThirdParty = true;
```

Comment:
Well, this `else` statement actually handles two cases at the same time: the case that a previously hidden third party becomes visible, and the case that a new third party that the first party has not linked to before and is not on the whitelist is present. Think about it as "new to the visualization", not "new to storage".